src/MSBuildBinLogQuery/Scan/Scanner.cs

```diff
@@ -0,0 +1,168 @@
+using System;
+using System.Text;
+using Microsoft.Build.Logging.Query.Token;
+
+namespace Microsoft.Build.Logging.Query.Scan
+{
+    public class Scanner
+    {
+        public Token.Token Token { get; private set; }
+
+        private readonly string _expression;
+        private int _index;
+        private char _char;
+
+        public Scanner(string expression)
+        {
+            _expression = expression ?? throw new ArgumentNullException();
+            _index = 0;
+
+            ReadNextCharacter();
+            ReadNextToken();
+        }
+
+        public void ReadNextToken()
+        {
+            SkipWhiteSpace();
+
+            switch (_char)
+            {
+                case '\0':
+                    Token = new EofToken();
```

Comment:
I would suggest making all of these that are effectively singletons by their value semantics actual singletons in memory. In other words, make the constructors private and have EofToken.Instance, etc.

Then you could restrict the IEquatable to the things that have more than 1 value like StringToken. For the single value ones, the inherited reference equality semantics would then be fine.

This will save on boilerplate in the token classes, and save on heap allocations during scanning.

Reply:
This is a great idea

Reply:
Done.